pinot-core/src/main/java/org/apache/pinot/core/plan/TransformPlanNode.java

```diff
@@ -43,76 +39,70 @@
+      Set<TransformExpressionTree> expressionsToPlan) {
     _segmentName = indexSegment.getSegmentName();
-    extractColumnsAndTransforms(brokerRequest, indexSegment);
-    _projectionPlanNode = new ProjectionPlanNode(indexSegment, _projectionColumns,
+
+    setMaxDocsForSelection(brokerRequest);
+    Set<String> projectionColumns = new HashSet<>();
+    extractProjectionColumns(expressionsToPlan, projectionColumns);
+
+    _expressions = expressionsToPlan;
+    _projectionPlanNode = new ProjectionPlanNode(indexSegment, projectionColumns,
         new DocIdSetPlanNode(indexSegment, brokerRequest, _maxDocPerNextCall));
   }
 
+  private void extractProjectionColumns(Set<TransformExpressionTree> expressionsToPlan, Set<String> projectionColumns) {
+    for (TransformExpressionTree expression : expressionsToPlan) {
+      extractProjectionColumns(expression, projectionColumns);
+    }
+  }
+
+  private void extractProjectionColumns(TransformExpressionTree expression, Set<String> projectionColumns) {
+    TransformExpressionTree.ExpressionType expressionType = expression.getExpressionType();
+    switch (expressionType) {
+      case FUNCTION:
+        for (TransformExpressionTree child : expression.getChildren()) {
+          extractProjectionColumns(child, projectionColumns);
+        }
+        break;
+
+      case IDENTIFIER:
+        projectionColumns.add(expression.getValue());
+        break;
+
+      case LITERAL:
+        // Do nothing.
+        break;
+
+      default:
+        throw new UnsupportedOperationException("Unsupported expression type: " + expressionType);
+    }
+  }
+
   /**
-   * Helper method to extract projection columns and transform expressions from the given broker request.
+   * Helper method to set the max number of docs to return for selection queries
    */
-  private void extractColumnsAndTransforms(BrokerRequest brokerRequest, IndexSegment indexSegment) {
-    Set<String> columns = new HashSet<>();
-    if (brokerRequest.isSetAggregationsInfo()) {
-      // Extract aggregation expressions
-      for (AggregationInfo aggregationInfo : brokerRequest.getAggregationsInfo()) {
-        if (aggregationInfo.getAggregationType().equalsIgnoreCase(AggregationFunctionType.DISTINCT.getName())) {
-          // 'DISTINCT(col1, col2 ...)' is modeled as one single aggregation function
-          List<String> distinctColumns = AggregationFunctionUtils.getAggregationExpressions(aggregationInfo);
-          columns.addAll(distinctColumns);
-        } else if (!aggregationInfo.getAggregationType().equalsIgnoreCase(AggregationFunctionType.COUNT.getName())) {
-          columns.addAll(AggregationFunctionUtils.getAggregationExpressions(aggregationInfo));
-        }
-      }
-      // Extract group-by expressions
-      if (brokerRequest.isSetGroupBy()) {
-        columns.addAll(brokerRequest.getGroupBy().getExpressions());
-      }
-    } else {
+  private void setMaxDocsForSelection(BrokerRequest brokerRequest) {
```

Comment:
This logic should also be handle in the upper level (SelectionPlanNode) and passed to this class

Reply:
Yeah I thought so too. But upper levels can be aggregation as well, which will now have this field leaked. For now I tend to keep it here (as was the case before) until the full cleanup happens.